Rubberduck.Parsing/Symbols/DeclarationFinder.cs

```diff
@@ -176,62 +182,59 @@ private void InitializeLazyCollections()
                     )
                 ).ToList();
 
-            var implementingNames = implementsInstructions.SelectMany(item =>
-                    Members(item.IdentifierReference.Declaration.QualifiedName.QualifiedModuleName)
-                        .Where(member => member.DeclarationType.HasFlag(DeclarationType.Member))
-                        .Select(member => item.IdentifierReference.Declaration.IdentifierName + "_" + member.IdentifierName)
-                        )
-                    .ToHashSet();
-
-            var implementableMembers = implementsInstructions.Select(item =>
-                new
-                {
-                    item.Context,
-                    Members = Members(item.IdentifierReference.QualifiedModuleName)
-                        .Where(implementingTypeMember => implementingNames.Contains(implementingTypeMember.IdentifierName))
-                        .ToList()
-                });
+            var output = new Dictionary<Tuple<VBAParser.ImplementsStmtContext, Declaration>, List<ModuleBodyElementDeclaration>>();
+            foreach (var impl in implementsInstructions)
+            {
+                output.Add(
+                    new Tuple<VBAParser.ImplementsStmtContext, Declaration>(impl.Context,
+                        impl.IdentifierReference.ParentScoping),
+                    ((ClassModuleDeclaration) (impl.IdentifierReference.ParentScoping)).Members.Where(item =>
+                        item is ModuleBodyElementDeclaration member && ReferenceEquals(member.InterfaceImplemented,
+                            impl.IdentifierReference.Declaration)).Cast<ModuleBodyElementDeclaration>().ToList());
+            }
 
-            return implementableMembers.ToDictionary(item => item.Context, item => item.Members);
+            return output;
         }
 
-        private IDictionary<Declaration, List<Declaration>> FindAllIinterfaceMembersByModule()
+        private Dictionary<ClassModuleDeclaration, List<ClassModuleDeclaration>> FindAllImplementionsByInterface()
         {
-            var implementsInstructions = UserDeclarations(DeclarationType.ClassModule)
-                .SelectMany(cls => cls.References
-                    .Where(reference => reference.Context is VBAParser.ImplementsStmtContext || reference.Context.IsDescendentOf<VBAParser.ImplementsStmtContext>())
-                    .Select(reference =>
-                        new
-                        {
-                            IdentifierReference = reference,
-                            Context = reference.Context is VBAParser.ImplementsStmtContext ?
-                                (VBAParser.ImplementsStmtContext)reference.Context
-                                : reference.Context.GetAncestor<VBAParser.ImplementsStmtContext>()
-                        }
-                    )
-                );
+            return UserDeclarations(DeclarationType.ClassModule)
+                .Cast<ClassModuleDeclaration>()
```

Comment:
Do we want to use `OfType<T>` instead to avoid throwing if something has the declaration type but does not derive from `ClassModuleDeclaration` or do we throw intentionally in that case?

Reply:
As long as it is intentional, I am OK with a hard cast.